exonum-java-binding/doc/Migration_guide_0.9.md

```diff
@@ -0,0 +1,183 @@
+The format of the metadata has not changed since 0.6.0.
+
+Artifacts are _deployed_ to the Exonum blockchain network; and then used to instantiate
+_service instances_. A single artifact can be used to create many instances. Each instance
+has a _name_ (ex `Service#getName`) and a numeric _identifier_ (ex `Service#getId`).
+An instance name is specified by the network administrators during service instantiation;
+an id is assigned by the framework automatically and can be [queried][service-id-lc-operation]
+using the light client.
+
+[service-id-lc-operation]: todo
+
+## Update the Service
+
+The [`Service`][service] implementation must be updated to support multiple instances;
+new initialization and reconfiguration procedures.
+
+1. Remove the hardcoded service id and name. Add a constructor parameter 
+`ServiceInstanceSpec instanceSpec` which provides instance name
+and id.
+    - If your service extends
+    [`com.exonum.binding.core.service.AbstractService`][abstract-service] —
+    pass the `instanceSpec` to the superclass constructor and use provided `AbstractService#getId`,
+    `getName`, and `getInstanceSpec` methods
+    - If it implements `Service` directly — save the needed information to fields.
+
+```java
+public final class FooService extends AbstractService {
+  @Inject
+  public FooService(ServiceInstanceSpec instanceSpec) {
+    super(instanceSpec);
+  }
+}
+```
+
+2. If the service requires initialization — migrate to the new 
+[initialization mechanism][service-initialize]. 
+
+3. If the service needs reconfiguration support: implement [`Configurable`][configurable] interface.
+Reconfiguration shall be performed through the new [supervisor][supervisor] service which
+replaces the configuration service.
+
+[service]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/service/Service.html
+[service-initialize]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/service/Service.html#initialize-com.exonum.binding.core.storage.database.Fork-com.exonum.binding.core.service.Configuration-
+[configurable]: https://exonum.com/doc/api/java-binding/0.9.0-rc1/com/exonum/binding/core/service/Configurable.html
+<!-- todo: Check the anchor when the docs land -->
+[supervisor]: https://exonum.com/doc/version/0.13-rc1/get-started/java-binding/#deploy-and-start-service
+
+### Update the Schema
+
+#### Isolate Collections via Namespaces
+
+Use a unique for each service instance namespace for service collections (e.g., a service name,
+or an id).
+
+#### Update ProofMaps
+
+The default [`ProofMap`][proof-map] implementation has been changed to hash user keys to produce an internal key.
+That allows _any_ type to be used as a key; but adds an extra hashing operation for every insert
+operation and changes how proofs need to be verified.
+
+In some cases (rarer than is usually considered), it might be needed to **not** hash the keys
```

Comment:
```suggestion
In some rare cases it is advisable **not** to hash the keys
```

Reply:
I am not sure about 'advisable' because the only positive effect of that is slightly higher performance; but to use it correctly a user must ensure that the referenced restrictions are met (they are here, till the link is fixed: https://github.com/exonum/exonum-java-binding/blob/4295cf898af1c64d40c50480b3135ed53c5f1339/exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/indices/ProofMapIndexProxy.java#L58-L69)

If they are met — it is better to use non-hashing variant.